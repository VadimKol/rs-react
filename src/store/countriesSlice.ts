import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { countries } from '@/common/countries';

import { type RootState } from './store';

export const countriesLice = createSlice({
  name: 'countries',
  initialState: countries,
  reducers: {
    addCountry: (state, action: PayloadAction<string>) => {
      state.push(action.payload);
    },
  },
});

export const { addCountry } = countriesLice.actions;

export const selectCountries = (state: RootState): string[] => state.countries;

export const useCountries = (): string[] => useSelector(selectCountries);
