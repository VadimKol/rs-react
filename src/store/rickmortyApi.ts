import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
import { getCharacter, getCharacters } from 'rickmortyapi';

export const rickmortyApi = createApi({
  reducerPath: 'rickmortyApi',
  baseQuery: fakeBaseQuery(),
  endpoints: (builder) => ({
    getCharacters: builder.query({
      queryFn: async ({ page, character }: { page: number; character: { name: string } }) => {
        try {
          const { data, status, statusMessage } = await getCharacters({ page, name: character.name });

          if (status === 200) {
            return { data: { characters: data.results || [], totalPages: data.info?.pages || 0 } };
          }
          if (status === 404) {
            return { data: { characters: [], totalPages: 0 } };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
    getCharacter: builder.query({
      queryFn: async (id: number) => {
        try {
          const { data, status, statusMessage } = await getCharacter(id);

          if (status === 200) {
            return { data };
          }
          if (status === 404) {
            return { data: null };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
  }),
});

export const { useGetCharactersQuery, useGetCharacterQuery } = rickmortyApi;
