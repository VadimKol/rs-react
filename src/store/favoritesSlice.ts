/* import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import { type Character } from 'rickmortyapi';

import { type RootState } from './store';

const initialState: Character[] = [];

export const favoritesSLice = createSlice({
  name: 'favorites',
  initialState,
  reducers: {
    addToFavorites: (state, action: PayloadAction<Character>) => {
      state.push(action.payload);
    },
    removeFromFavorites: (state, action: PayloadAction<Character>) =>
      state.filter((character) => character.id !== action.payload.id),
    clearFavorites: () => [],
  },
});

export const { addToFavorites, removeFromFavorites, clearFavorites } = favoritesSLice.actions;

const selectFavorites = (state: RootState): Character[] => state.favorites;

const isFavorite = (state: RootState, character?: Character): boolean =>
  state.favorites.some((favorite) => favorite.id === character?.id);

export const useFavorites = (): Character[] => useSelector(selectFavorites);

export const useIsFavorite = (character?: Character): boolean =>
  useSelector((state: RootState) => isFavorite(state, character));
 */
