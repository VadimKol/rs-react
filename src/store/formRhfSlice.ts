import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

const initialState: FormInfo = {
  name: null,
  age: null,
  email: null,
  password: null,
  confirmPassword: null,
  gender: null,
  country: null,
  tc: null,
  image: null,
  strength: '0',
};

export const formRhfSLice = createSlice({
  name: 'formRhf',
  initialState,
  reducers: {
    updateFormRhf: (_, action: PayloadAction<FormInfo>) => action.payload,
  },
});

export const { updateFormRhf } = formRhfSLice.actions;

const selectFormRhfInfo = (state: RootState): FormInfo => state.formRhf;

export const useFormRhfInfo = (): FormInfo => useSelector(selectFormRhfInfo);
