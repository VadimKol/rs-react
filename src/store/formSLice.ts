import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

export const initialState: FormInfo[] = [];

export const formSLice = createSlice({
  name: 'form',
  initialState,
  reducers: {
    updateForm: (state, action: PayloadAction<FormInfo>) => {
      state.push(action.payload);
    },
  },
});

export const { updateForm } = formSLice.actions;

const selectFormInfo = (state: RootState): FormInfo[] => state.form;

export const useFormInfo = (): FormInfo[] => useSelector(selectFormInfo);
