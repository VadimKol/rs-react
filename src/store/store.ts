import { configureStore } from '@reduxjs/toolkit';

import { formSLice } from './formSLice';

export const store = configureStore({ reducer: { form: formSLice.reducer } });

export type RootState = ReturnType<typeof store.getState>;
