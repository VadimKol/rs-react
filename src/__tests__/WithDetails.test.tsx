import { render, waitFor } from '@testing-library/react';
import mockRouter from 'next-router-mock';
import { Provider } from 'react-redux';

import WithDetails from '@/pages/character/[id]';
import { makeStore } from '@/store/store';

import { character, characters } from './__mocks__/data';

global.URL.createObjectURL = jest.fn();
global.URL.revokeObjectURL = jest.fn();

jest.mock('@/store/rickmortyApi', () => ({
  ...jest.requireActual<object>('@/store/rickmortyApi'),
  useGetCharactersQuery: jest.fn(() => ({ data: { characters, totalPages: 1 }, isError: false, error: undefined })),
  useGetCharacterQuery: jest.fn(() => ({ data: character, isFetching: false, isError: false, error: undefined })),
}));

const store = makeStore();

describe('Main page with details', () => {
  it('renders correctly', async () => {
    mockRouter.push({ pathname: '/', query: { page: '1', search: 'Rick', id: '1' } });

    const { container } = render(
      <Provider store={store}>
        <WithDetails />
      </Provider>,
    );

    await waitFor(() => expect(container).toMatchSnapshot());
  });
});
