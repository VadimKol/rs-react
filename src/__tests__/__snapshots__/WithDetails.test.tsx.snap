// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page with details renders correctly 1`] = `
<div>
  <main
    class="main"
    role="presentation"
  >
    <section>
      <form
        class="search_form"
      >
        <input
          class="search"
          id="search"
          placeholder="Search..."
          type="text"
          value="Rick"
        />
        <button
          aria-label="Search-button"
          class="search_button"
          type="submit"
        />
      </form>
    </section>
    <section
      class="results"
    >
      <div
        class="character"
      >
        <div
          class="results"
        >
          <ul
            class="characters"
            role="presentation"
          >
            <li
              class="characters_item"
            >
              <div
                class="card"
              >
                <div
                  class="image_container"
                >
                  <div
                    class="loader"
                  />
                  <img
                    alt="Character"
                    class="hidden"
                    data-nimg="1"
                    decoding="async"
                    height="300"
                    loading="lazy"
                    src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 2x"
                    style="color: transparent;"
                    width="300"
                  />
                </div>
                <div
                  class="text_container"
                >
                  <a
                    class="card_link"
                    href="/character/1?page=1&search=Rick"
                  >
                    <h2
                      class="card_title"
                    >
                      Rick
                    </h2>
                  </a>
                  <button
                    aria-label="Add to favorites"
                    class="favorite"
                    name="favorite"
                    type="button"
                    value="false"
                  >
                    ☆
                  </button>
                </div>
              </div>
            </li>
            <li
              class="characters_item"
            >
              <div
                class="card"
              >
                <div
                  class="image_container"
                >
                  <div
                    class="loader"
                  />
                  <img
                    alt="Character"
                    class="hidden"
                    data-nimg="1"
                    decoding="async"
                    height="300"
                    loading="lazy"
                    src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
                    srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
                    style="color: transparent;"
                    width="300"
                  />
                </div>
                <div
                  class="text_container"
                >
                  <a
                    class="card_link"
                    href="/character/2?page=1&search=Rick"
                  >
                    <h2
                      class="card_title"
                    >
                      Morty
                    </h2>
                  </a>
                  <button
                    aria-label="Add to favorites"
                    class="favorite"
                    name="favorite"
                    type="button"
                    value="false"
                  >
                    ☆
                  </button>
                </div>
              </div>
            </li>
          </ul>
          <div
            class="pagination"
          >
            <div
              class="pagination_box"
            >
              <div
                class="pages"
              >
                <span>
                  1
                </span>
                -
                <span>
                  1
                </span>
              </div>
            </div>
          </div>
        </div>
        <section
          class="detailed_Card"
        >
          <div
            class="card"
          >
            <div
              class="image_container"
            >
              <div
                class="loader"
              />
              <img
                alt="Character"
                class="hidden"
                data-nimg="1"
                decoding="async"
                height="300"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 2x"
                style="color: transparent;"
                width="300"
              />
            </div>
            <div
              class="text_container"
            >
              <h2
                class="card_title"
              >
                Rick
              </h2>
              <ul
                class="desc"
              >
                <li
                  class="desc_item"
                >
                  Species: 
                </li>
                <li
                  class="desc_item"
                >
                  Gender: Male
                </li>
                <li
                  class="desc_item"
                >
                  Status: Alive
                </li>
                <li
                  class="desc_item"
                >
                  Origin: 
                </li>
                <li
                  class="desc_item"
                >
                  Location: 
                </li>
              </ul>
              <p
                class="episodes"
              >
                Episodes: 
              </p>
              <button
                aria-label="Add to favorites"
                class="favorite"
                name="favorite"
                type="button"
                value="false"
              >
                ☆
              </button>
            </div>
          </div>
          <button
            aria-label="Button"
            class="button close"
            id=""
            type="button"
          >
            Close
          </button>
        </section>
      </div>
    </section>
  </main>
</div>
`;
